### PR TITLE
i1870 filter /scan/unspent boxes by inclusion height efficiently

### PR DESCRIPTION
**Pull request by pragmaxim**
closes https://github.com/ergoplatform/ergo/issues/1870

@kushti I need more time for this one, the KV store where the keys are composed of multiple "navigation" byte sequences are very hard to reason about.

### CODE REVIEW

**Review by kushti — CHANGES_REQUESTED** [review]
It seems spent boxes are included into results.

Submitted failing test case.

I think you need to pass height into

```
  private def firstScanBoxSpaceKey(scanId: ScanId): Array[Byte] =
    composeKey(UnspentIndexPrefix, scanId, 0)

  private def lastScanBoxSpaceKey(scanId: ScanId): Array[Byte] =
    composeKey(UnspentIndexPrefix, scanId, -1)
```

**Comment by pragmaxim** [comment]
@kushti The problem I encountered was backward compatibility : 
https://github.com/ergoplatform/ergo/blob/39fab7039f02278cffdfbb6b2632982254bb0c80/src/main/scala/org/ergoplatform/nodeView/wallet/persistence/WalletRegistry.scala#L519-L520

I'd have to put `height` here like this `composeKeyWithId(prefix, scanId, height, trackedBox.box.id)` but it would no longer be backward compatible, right?

**Comment by kushti** [comment]
@pragmaxim oh yes, indeed my proposed solution is breaking backwards compatibility. With your implementation, reading db is not efficient for big ranges. Maybe we need to use new method for small ranges, and old one otherwise ?

**Comment by pragmaxim** [comment]
@kushti on the contrary, I believe that this will be faster as we called `sortByHeight` on unordered boxes and now we load the boxes already sorted by height.

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletService.scala [hunk 1]
```diff
@@ -405,16 +407,21 @@ class ErgoWalletServiceImpl(override val ergoSettings: ErgoSettings) extends Erg
...
   }
 
-  override def getScanUnspentBoxes(state: ErgoWalletState, scanId: ScanId, considerUnconfirmed: Boolean): Seq[WalletBox] = {
+  override def getScanUnspentBoxes(state: ErgoWalletState, scanId: ScanId, considerUnconfirmed: Boolean, minHeight: Int, maxHeight: Int): Seq[WalletBox] = {
     val unconfirmed = if (considerUnconfirmed) {
-      state.offChainRegistry.offChainBoxes.filter(_.scans.contains(scanId))
+      state.offChainRegistry.offChainBoxes
+        .filter { box =>
+          box.scans.contains(scanId) &&
+          box.inclusionHeightOpt.getOrElse(0) >= minHeight  &&
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletService.scala:415]: for unconfirmed boxes, inclusion height is always defined

I think you need to return unconfirmed with old filter, and now only if maxHeight is not defined by a user

**Review by kushti — CHANGES_REQUESTED** [review]

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletService.scala [hunk 2]
```diff
@@ -405,16 +407,21 @@ class ErgoWalletServiceImpl(override val ergoSettings: ErgoSettings) extends Erg
     boxes.map(tb => WalletBox(tb, currentHeight)).sortBy(_.trackedBox.inclusionHeightOpt)
   }
 
-  override def getScanUnspentBoxes(state: ErgoWalletState, scanId: ScanId, considerUnconfirmed: Boolean): Seq[WalletBox] = {
+  override def getScanUnspentBoxes(state: ErgoWalletState, scanId: ScanId, considerUnconfirmed: Boolean, minHeight: Int, maxHeight: Int): Seq[WalletBox] = {
```
[the remark this answers is not in the document]
Comment (pragmaxim) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletService.scala:410]: It is `override `, all the `ErgoWalletService` implementations are documented in the interface methods.

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletService.scala (hunk 1 above)
Comment (pragmaxim) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletService.scala:415]: I don't understand this comment, old filter was `_.scans.contains(scanId)` and now we also want to filter out boxes unconfirmed boxes that are out of height range :thinking:

Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletService.scala:415]: Unconfirmed boxes do not have inclusion height, they are not in the chain yet

Comment (pragmaxim) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletService.scala]: I see, we could think about it also in a way that `considerUnconfirmed` overrules the height range ...

Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletService.scala]: I think we need to allow considerUnconfirmed to be set only if there's no limit for maxHeight.

**Review by kushti — APPROVED** [review]

**Pull request merged**